vision/why-compile-time-evaluation.md

```diff
@@ -0,0 +1,44 @@
+# Why compile time evaluation?
+
+This section is, in many ways, the most important. It aims to clarify the reasoning about any future decisions about const generics and const evaluation.
+
+## Why do we want to evaluate things at compile time?
+
+People already precompute certain data structures, lookup tables, and so on, at compile time.
+For example the crate `regex` could be able to compile a regex at compile time,
+improving the performance when using `regex` and even removing the need for runtime allocations. 
+
+Without good const evaluation support, is often easier to just use a `build.rs` file for this or compute it externally and generate the result outside of the ordinary compilation process. If that isn't possible or sensible, one often just requires the computation to happen at runtime, negatively impacting startup performance.
+
+Computing things outside without using const eval makes it at lot more cumbersome to use and
+prevents the input for that computation to simply be a generic or function parameter.
```

Comment:
It also requires people to essentially learn to generate code, instead of just writing the code they would write at runtime, and knowing that the compiler will make sure there is no runtime cost for it.